// @thirdweb-dev/auth/dist/thirdweb-dev-auth.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var auth = require('./auth-76921cbf.cjs.prod.js');
require('zod');
require('uuid');
require('ethers');



exports.AccountTypeSchema = auth.AccountTypeSchema;
exports.AddressSchema = auth.AddressSchema;
exports.AuthenticateOptionsSchema = auth.AuthenticateOptionsSchema;
exports.AuthenticationPayloadDataSchema = auth.AuthenticationPayloadDataSchema;
exports.AuthenticationPayloadSchema = auth.AuthenticationPayloadSchema;
exports.GenerateOptionsSchema = auth.GenerateOptionsSchema;
exports.JsonSchema = auth.JsonSchema;
exports.LoginOptionsSchema = auth.LoginOptionsSchema;
exports.LoginPayloadDataSchema = auth.LoginPayloadDataSchema;
exports.LoginPayloadOutputSchema = auth.LoginPayloadOutputSchema;
exports.LoginPayloadSchema = auth.LoginPayloadSchema;
exports.RawDateSchema = auth.RawDateSchema;
exports.RefreshOptionsSchema = auth.RefreshOptionsSchema;
exports.ThirdwebAuth = auth.ThirdwebAuth;
exports.VerifyOptionsSchema = auth.VerifyOptionsSchema;
exports.authenticateJWT = auth.authenticateJWT;
exports.buildAndSignLoginPayload = auth.buildAndSignLoginPayload;
exports.buildJWT = auth.buildJWT;
exports.buildLoginPayload = auth.buildLoginPayload;
exports.generateJWT = auth.generateJWT;
exports.parseJWT = auth.parseJWT;
exports.refreshJWT = auth.refreshJWT;
exports.signLoginPayload = auth.signLoginPayload;
exports.verifyLoginPayload = auth.verifyLoginPayload;
