// @web3uikit/core/dist/ButtonColored.js
import { C as Ce, g as getShade, c as color, s as styled, B as ButtonBase, j as jsx } from "./index.js";
import "react";
import "react-router-dom";
const coloredShades = Ce`
    :after {
        background-color: ${getShade("light", 90)};
    }

    :hover {
        :after {
            background-color: ${getShade("light", 70)};
        }
    }

    :active {
        :after {
            background-color: ${getShade("light", 50)};
        }
    }
`;
const coloredRed = Ce`
    background-color: ${color.red40};
    border-color: ${color.red40};
    color: ${color.red40};

    :focus {
        box-shadow: 0px 0px 0px 2px ${color.navy30};
    }

    svg {
        fill: ${color.red40};
    }

    ${coloredShades}
`;
const coloredGreen = Ce`
    background-color: ${color.mint40};
    border-color: ${color.mint40};
    color: ${color.mint40};

    :focus {
        box-shadow: 0px 0px 0px 2px ${color.navy30};
    }

    svg {
        fill: ${color.mint40};
    }

    ${coloredShades}
`;
const coloredBlue = Ce`
    background-color: ${color.navy40};
    border-color: ${color.navy40};
    color: ${color.navy40};

    :focus {
        box-shadow: 0px 0px 0px 2px ${color.navy30};
    }

    svg {
        fill: ${color.navy40};
    }

    ${coloredShades}
`;
const coloredYellow = Ce`
    background-color: ${color.yellow50};
    border-color: ${color.yellow50};
    color: ${color.yellow50};

    :focus {
        box-shadow: 0px 0px 0px 2px ${color.navy30};
    }

    svg {
        fill: ${color.yellow50};
    }

    ${coloredShades}
`;
const getColored = (color2) => {
  switch (color2) {
    case "red":
      return coloredRed;
    case "green":
      return coloredGreen;
    case "blue":
      return coloredBlue;
    case "yellow":
      return coloredYellow;
    default:
      return;
  }
};
const ButtonColoredStyled$1 = styled(ButtonBase)`
    :after {
        background-color: ${getShade("dark", 0)};
        content: '';
        display: block;
        height: 100%;
        left: 0;
        pointer-events: none;
        position: absolute;
        top: 0;
        transition: all 0.3s ease;
        width: 100%;
        z-index: 0;
        border-radius: 10px;
    }

    ${({
  color: color2
}) => color2 && getColored(color2)}
`;
var styles = {
  ButtonColoredStyled: ButtonColoredStyled$1
};
const {
  ButtonColoredStyled
} = styles;
const ButtonColored = ({
  color: color2,
  ...props
}) => /* @__PURE__ */ jsx(ButtonColoredStyled, {
  color: color2,
  ...props
});
export { ButtonColored as default };
