// @thirdweb-dev/wallets/evm/connectors/zerion/dist/thirdweb-dev-wallets-evm-connectors-zerion.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var evm_connectors_injected_dist_thirdwebDevWalletsEvmConnectorsInjected = require('../../injected/dist/thirdweb-dev-wallets-evm-connectors-injected.cjs.dev.js');
var assertWindowEthereum = require('../../../../dist/assertWindowEthereum-0709ee9e.cjs.dev.js');
require('../../../../dist/defineProperty-982e7acf.cjs.dev.js');
require('../../../../dist/classPrivateFieldSet-f0dc2a42.cjs.dev.js');
require('../../../../dist/WagmiConnector-6ff7d562.cjs.dev.js');
require('@thirdweb-dev/chains');
require('eventemitter3');
require('../../../../dist/errors-e4b137ea.cjs.dev.js');
require('../../../../dist/url-4b641c31.cjs.dev.js');
require('ethers');
require('../../../../dist/normalizeChainId-5da85f42.cjs.dev.js');

class ZerionConnector extends evm_connectors_injected_dist_thirdwebDevWalletsEvmConnectorsInjected.InjectedConnector {
  constructor(arg) {
    const defaultOptions = {
      name: "Zerion",
      getProvider() {
        function getReady(ethereum) {
          const isZerion = !!ethereum?.isZerion;
          if (!isZerion) {
            return;
          }
          return ethereum;
        }
        if (typeof window === "undefined") {
          return;
        }
        if (assertWindowEthereum.assertWindowEthereum(globalThis.window)) {
          if (globalThis.window.ethereum?.providers) {
            return globalThis.window.ethereum.providers.find(getReady);
          }
          return getReady(globalThis.window.ethereum);
        }
      }
    };
    const options = {
      ...defaultOptions,
      ...arg.options
    };
    super({
      chains: arg.chains,
      options,
      connectorStorage: arg.connectorStorage
    });
  }
}

exports.ZerionConnector = ZerionConnector;
