// @thirdweb-dev/wallets/evm/connectors/zerion/dist/thirdweb-dev-wallets-evm-connectors-zerion.esm.js
import { InjectedConnector } from '../../injected/dist/thirdweb-dev-wallets-evm-connectors-injected.esm.js';
import { a as assertWindowEthereum } from '../../../../dist/assertWindowEthereum-2bcf9787.esm.js';
import '../../../../dist/defineProperty-d067c271.esm.js';
import '../../../../dist/classPrivateFieldSet-e25e1cec.esm.js';
import '../../../../dist/WagmiConnector-6011bbb1.esm.js';
import '@thirdweb-dev/chains';
import 'eventemitter3';
import '../../../../dist/errors-b8627438.esm.js';
import '../../../../dist/url-0d129c6b.esm.js';
import 'ethers';
import '../../../../dist/normalizeChainId-abcb61d5.esm.js';

class ZerionConnector extends InjectedConnector {
  constructor(arg) {
    const defaultOptions = {
      name: "Zerion",
      getProvider() {
        function getReady(ethereum) {
          const isZerion = !!ethereum?.isZerion;
          if (!isZerion) {
            return;
          }
          return ethereum;
        }
        if (typeof window === "undefined") {
          return;
        }
        if (assertWindowEthereum(globalThis.window)) {
          if (globalThis.window.ethereum?.providers) {
            return globalThis.window.ethereum.providers.find(getReady);
          }
          return getReady(globalThis.window.ethereum);
        }
      }
    };
    const options = {
      ...defaultOptions,
      ...arg.options
    };
    super({
      chains: arg.chains,
      options,
      connectorStorage: arg.connectorStorage
    });
  }
}

export { ZerionConnector };
