// @thirdweb-dev/wallets/evm/connectors/zerion/dist/thirdweb-dev-wallets-evm-connectors-zerion.browser.esm.js
import { InjectedConnector } from '../../injected/dist/thirdweb-dev-wallets-evm-connectors-injected.browser.esm.js';
import { a as assertWindowEthereum } from '../../../../dist/assertWindowEthereum-88295886.browser.esm.js';
import '../../../../dist/defineProperty-d7c057bf.browser.esm.js';
import '../../../../dist/classPrivateFieldSet-a5db7c83.browser.esm.js';
import '../../../../dist/WagmiConnector-2f14002d.browser.esm.js';
import '@thirdweb-dev/chains';
import 'eventemitter3';
import '../../../../dist/errors-9f6736cb.browser.esm.js';
import '../../../../dist/url-a45219bd.browser.esm.js';
import 'ethers';
import '../../../../dist/normalizeChainId-1fb9aedf.browser.esm.js';

class ZerionConnector extends InjectedConnector {
  constructor(arg) {
    const defaultOptions = {
      name: "Zerion",
      getProvider() {
        function getReady(ethereum) {
          const isZerion = !!ethereum?.isZerion;
          if (!isZerion) {
            return;
          }
          return ethereum;
        }
        if (assertWindowEthereum(globalThis.window)) {
          if (globalThis.window.ethereum?.providers) {
            return globalThis.window.ethereum.providers.find(getReady);
          }
          return getReady(globalThis.window.ethereum);
        }
      }
    };
    const options = {
      ...defaultOptions,
      ...arg.options
    };
    super({
      chains: arg.chains,
      options,
      connectorStorage: arg.connectorStorage
    });
  }
}

export { ZerionConnector };
