// @thirdweb-dev/auth/express/dist/thirdweb-dev-auth-express.esm.js
import { n as THIRDWEB_AUTH_ACTIVE_ACCOUNT_COOKIE, o as THIRDWEB_AUTH_TOKEN_COOKIE_PREFIX, l as LoginPayloadOutputSchema, q as THIRDWEB_AUTH_DEFAULT_REFRESH_INTERVAL_IN_SECONDS, t as THIRDWEB_AUTH_DEFAULT_TOKEN_DURATION_IN_SECONDS, T as ThirdwebAuth$1 } from '../../dist/auth-3e6bcaed.esm.js';
import { z } from 'zod';
import { serialize } from 'cookie';
import cookieParser from 'cookie-parser';
import express from 'express';
import 'uuid';
import 'ethers';

function getCookie(req, cookie) {
  if (typeof req.cookies.get === "function") {
    return req.cookies.get(cookie);
  }
  return req.cookies[cookie];
}
function getActiveCookie(req) {
  if (!req.cookies) {
    return undefined;
  }
  const activeAccount = getCookie(req, THIRDWEB_AUTH_ACTIVE_ACCOUNT_COOKIE);
  if (activeAccount) {
    return `${THIRDWEB_AUTH_TOKEN_COOKIE_PREFIX}_${activeAccount}`;
  }

  // If active account is not present, then use the old default
  return THIRDWEB_AUTH_TOKEN_COOKIE_PREFIX;
}

/**
 * @internal
 * @param req
 * @returns
 */
function getToken(req) {
  if (req.headers["authorization"]) {
    const authorizationHeader = req.headers["authorization"].split(" ");
    if (authorizationHeader?.length === 2) {
      return authorizationHeader[1];
    }
  }
  if (!req.cookies) {
    return undefined;
  }
  const activeCookie = getActiveCookie(req);
  if (!activeCookie) {
    return undefined;
  }
  return getCookie(req, activeCookie);
}
async function getUser(req, ctx) {
  const token = getToken(req);
  if (!token) {
    return null;
  }
  let authenticatedUser;
  try {
    authenticatedUser = await ctx.auth.authenticate(token, {
      validateTokenId: async tokenId => {
        if (ctx.authOptions?.validateTokenId) {
          await ctx.authOptions?.validateTokenId(tokenId);
        }
      }
    });
  } catch (err) {
    return null;
  }
  if (!ctx.callbacks?.onUser) {
    return authenticatedUser;
  }
  const data = await ctx.callbacks.onUser(authenticatedUser, req);
  if (!data) {
    return authenticatedUser;
  }
  return {
    ...authenticatedUser,
    data: data
  };
}

const PayloadBodySchema = z.object({
  address: z.string(),
  chainId: z.string().optional()
});
const ActiveBodySchema = z.object({
  address: z.string()
});
const LoginPayloadBodySchema = z.object({
  payload: LoginPayloadOutputSchema
});

async function handler$4(req, res, ctx) {
  if (req.method !== "POST") {
    return res.status(405).json({
      error: "Invalid method. Only POST supported."
    });
  }
  const parsedPayload = PayloadBodySchema.safeParse(req.body);
  if (!parsedPayload.success) {
    return res.status(400).json({
      error: "Please provide an address"
    });
  }

  // TODO: Add nonce generation + custom expiration + invalid before
  const payload = await ctx.auth.payload({
    address: parsedPayload.data.address,
    statement: ctx.authOptions?.statement,
    uri: ctx.authOptions?.uri,
    version: ctx.authOptions?.version,
    chainId: parsedPayload.data.chainId || ctx.authOptions?.chainId,
    resources: ctx.authOptions?.resources,
    expirationTime: ctx.authOptions?.loginPayloadDurationInSeconds ? new Date(Date.now() + 1000 * ctx.authOptions.loginPayloadDurationInSeconds) : undefined
  });
  return res.status(200).json({
    payload
  });
}

async function handler$3(req, res, ctx) {
  if (req.method !== "POST") {
    return res.status(405).json({
      error: "Method not allowed"
    });
  }
  const parsedPayload = LoginPayloadBodySchema.safeParse(req.body);

  // Get signed login payload from the frontend
  if (!parsedPayload.success) {
    return res.status(400).json({
      error: "Invalid login payload"
    });
  }
  const payload = parsedPayload.data.payload;
  const validateNonce = async nonce => {
    if (ctx.authOptions?.validateNonce) {
      await ctx.authOptions?.validateNonce(nonce);
    }
  };
  const getSession = async address => {
    if (ctx.callbacks?.onLogin) {
      return ctx.callbacks.onLogin(address, req);
    }
  };
  const expirationTime = ctx.authOptions?.tokenDurationInSeconds ? new Date(Date.now() + 1000 * ctx.authOptions.tokenDurationInSeconds) : undefined;
  const generateOptions = {
    verifyOptions: {
      statement: ctx.authOptions?.statement,
      uri: ctx.authOptions?.uri,
      version: ctx.authOptions?.version,
      chainId: ctx.authOptions?.chainId,
      validateNonce,
      resources: ctx.authOptions?.resources
    },
    expirationTime,
    session: getSession
  };
  let token;
  try {
    // Generate an access token with the SDK using the signed payload
    token = await ctx.auth.generate(payload, generateOptions);
  } catch (err) {
    if (err.message) {
      return res.status(400).json({
        error: err.message
      });
    } else if (typeof err === "string") {
      return res.status(400).json({
        error: err
      });
    } else {
      return res.status(400).json({
        error: "Invalid login payload"
      });
    }
  }
  if (ctx.callbacks?.onToken) {
    await ctx.callbacks.onToken(token, req);
  }
  const {
    payload: {
      exp
    }
  } = ctx.auth.parseToken(token);

  // Securely set httpOnly cookie on request to prevent XSS on frontend
  // And set path to / to enable thirdweb_auth_token usage on all endpoints
  res.setHeader("Set-Cookie", [serialize(`${THIRDWEB_AUTH_TOKEN_COOKIE_PREFIX}_${payload.payload.address}`, token, {
    domain: ctx.cookieOptions?.domain,
    path: ctx.cookieOptions?.path || "/",
    sameSite: ctx.cookieOptions?.sameSite || "none",
    expires: new Date(exp * 1000),
    httpOnly: true,
    secure: ctx.cookieOptions?.secure || true
  }), serialize(THIRDWEB_AUTH_ACTIVE_ACCOUNT_COOKIE, payload.payload.address, {
    domain: ctx.cookieOptions?.domain,
    path: ctx.cookieOptions?.path || "/",
    sameSite: ctx.cookieOptions?.sameSite || "none",
    expires: new Date(exp * 1000),
    httpOnly: true,
    secure: ctx.cookieOptions?.secure || true
  })]);

  // Send token in body and as cookie for frontend and backend use cases
  return res.status(200).json({
    token
  });
}

async function handler$2(req, res, ctx) {
  if (req.method !== "POST") {
    return res.status(405).json({
      error: "Invalid method. Only POST supported."
    });
  }
  const activeCookie = getActiveCookie(req);
  if (!activeCookie) {
    return res.status(400).json({
      error: "No logged in user to logout."
    });
  }
  if (ctx.callbacks?.onLogout) {
    const user = await getUser(req, ctx);
    if (user) {
      await ctx.callbacks.onLogout(user, req);
    }
  }

  // Set the access token to 'none' and expire in 5 seconds
  res.setHeader("Set-Cookie", serialize(activeCookie, "", {
    domain: ctx.cookieOptions?.domain,
    path: ctx.cookieOptions?.path || "/",
    sameSite: ctx.cookieOptions?.sameSite || "none",
    expires: new Date(Date.now() + 5 * 1000),
    httpOnly: true,
    secure: ctx.cookieOptions?.secure || true
  }));
  return res.status(200).json({
    message: "Successfully logged out"
  });
}

async function handler$1(req, res, ctx) {
  if (req.method !== "GET") {
    return res.status(400).json({
      error: "Invalid method. Only GET supported."
    });
  }
  const user = await getUser(req, ctx);

  // Importantly, make sure the user was actually logged in before refreshing
  if (user) {
    const token = getToken(req);
    if (token) {
      const payload = ctx.auth.parseToken(token);
      const refreshDate = ctx.authOptions?.refreshIntervalInSeconds ? new Date(payload.payload.iat * 1000 + ctx.authOptions.refreshIntervalInSeconds * 1000) : new Date(payload.payload.iat * 1000 + THIRDWEB_AUTH_DEFAULT_REFRESH_INTERVAL_IN_SECONDS * 1000);
      if (new Date() > refreshDate) {
        const expirationTime = ctx.authOptions?.tokenDurationInSeconds ? new Date(Date.now() + 1000 * ctx.authOptions.tokenDurationInSeconds) : undefined;
        const refreshedToken = await ctx.auth.refresh(token, expirationTime);
        const refreshedPayload = ctx.auth.parseToken(refreshedToken);
        res.setHeader("Set-Cookie", [serialize(`${THIRDWEB_AUTH_TOKEN_COOKIE_PREFIX}_${user.address}`, refreshedToken, {
          domain: ctx.cookieOptions?.domain,
          path: ctx.cookieOptions?.path || "/",
          sameSite: ctx.cookieOptions?.sameSite || "none",
          expires: new Date(refreshedPayload.payload.exp * 1000),
          httpOnly: true,
          secure: ctx.cookieOptions?.secure || true
        }), serialize(THIRDWEB_AUTH_ACTIVE_ACCOUNT_COOKIE, user.address, {
          domain: ctx.cookieOptions?.domain,
          path: ctx.cookieOptions?.path || "/",
          sameSite: ctx.cookieOptions?.sameSite || "none",
          expires: new Date(refreshedPayload.payload.exp * 1000),
          httpOnly: true,
          secure: ctx.cookieOptions?.secure || true
        })]);
      }
    }
  }
  return res.status(200).json(user);
}

async function handler(req, res, ctx) {
  if (req.method !== "POST") {
    return res.status(405).json({
      error: "Invalid method. Only POST supported."
    });
  }
  const parsedPayload = ActiveBodySchema.safeParse(req.body);
  if (!parsedPayload.success) {
    return res.status(400).json({
      error: "Please provide an address"
    });
  }
  let cookieExpiration;
  const cookie = getCookie(req, `${THIRDWEB_AUTH_TOKEN_COOKIE_PREFIX}_${parsedPayload.data.address}`);
  if (cookie) {
    // If the new account is already logged in, get the expiration time from the cookie
    const {
      payload: {
        exp
      }
    } = ctx.auth.parseToken(cookie);
    cookieExpiration = new Date(exp * 1000);
  } else if (ctx.authOptions?.tokenDurationInSeconds) {
    // Otherwise, if we have a token duration in seconds, set it to that
    cookieExpiration = new Date(Date.now() + 1000 * ctx.authOptions.tokenDurationInSeconds);
  } else {
    // Otherwise, just default to 24 hours
    cookieExpiration = new Date(Date.now() + 1000 * THIRDWEB_AUTH_DEFAULT_TOKEN_DURATION_IN_SECONDS);
  }
  res.setHeader("Set-Cookie", [serialize(THIRDWEB_AUTH_ACTIVE_ACCOUNT_COOKIE, parsedPayload.data.address, {
    domain: ctx.cookieOptions?.domain,
    path: ctx.cookieOptions?.path || "/",
    sameSite: ctx.cookieOptions?.sameSite || "none",
    expires: cookieExpiration,
    httpOnly: true,
    secure: ctx.cookieOptions?.secure || true
  })]);
  return res.status(200).end();
}

const asyncHandler = fn => function () {
  var _ref;
  const fnReturn = fn(...arguments);
  const next = (_ref = arguments.length - 1, _ref < 0 || arguments.length <= _ref ? undefined : arguments[_ref]);
  return Promise.resolve(fnReturn).catch(next);
};
function ThirdwebAuth(cfg) {
  const ctx = {
    ...cfg,
    auth: new ThirdwebAuth$1(cfg.wallet, cfg.domain)
  };
  const router = express.Router();
  const cookieMiddleware = cookieParser();
  router.use(express.json());
  router.use(cookieMiddleware);
  router.post("/payload", asyncHandler((req, res) => handler$4(req, res, ctx)));
  router.post("/login", asyncHandler((req, res) => handler$3(req, res, ctx)));
  router.get("/user", asyncHandler((req, res) => handler$1(req, res, ctx)));
  router.post("/logout", asyncHandler((req, res) => handler$2(req, res, ctx)));
  router.post("/switch-account", asyncHandler((req, res) => handler(req, res, ctx)));
  return {
    authRouter: router,
    authMiddleware: cookieMiddleware,
    getUser: req => {
      return getUser(req, ctx);
    }
  };
}

export { ActiveBodySchema, LoginPayloadBodySchema, PayloadBodySchema, ThirdwebAuth, getToken };
