// @thirdweb-dev/react/evm/dist/thirdweb-dev-react-evm.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var formElements = require('../../dist/formElements-69bdb39b.cjs.prod.js');
var useSafe = require('../../dist/useSafe-6e80e3d3.cjs.prod.js');
var useMagic = require('../../dist/useMagic-f2fd7684.cjs.prod.js');
var oneKeyWallet = require('../../dist/oneKeyWallet-ad515296.cjs.prod.js');
var reactCore = require('@thirdweb-dev/react-core');
var wallets = require('@thirdweb-dev/wallets');
require('react/jsx-runtime');
require('@radix-ui/react-icons');
require('@emotion/react');
require('react');
require('@radix-ui/colors');
require('@emotion/styled');
require('@tanstack/react-query');
require('@radix-ui/react-popover');
require('copy-to-clipboard');
require('../../dist/Tooltip-adf19cd9.cjs.prod.js');
require('@radix-ui/react-tooltip');
require('../../dist/safeChainSlug-c20b69eb.cjs.prod.js');
require('detect-browser');
require('qrcode');
require('@radix-ui/react-tabs');
require('fuse.js');
require('@thirdweb-dev/chains');
require('ethers');
require('@thirdweb-dev/sdk');
require('tiny-invariant');
require('@radix-ui/react-dialog');
require('@radix-ui/react-focus-scope');



exports.useIsWalletModalOpen = formElements.useIsWalletModalOpen;
exports.useSetIsWalletModalOpen = formElements.useSetIsWalletModalOpen;
exports.useSetWalletModalConfig = formElements.useSetWalletModalConfig;
exports.useSafe = useSafe.useSafe;
exports.useMagic = useMagic.useMagic;
exports.ConnectEmbed = oneKeyWallet.ConnectEmbed;
exports.ConnectModalInline = oneKeyWallet.ConnectModalInline;
exports.ConnectWallet = oneKeyWallet.ConnectWallet;
exports.MediaRenderer = oneKeyWallet.MediaRenderer;
exports.NetworkSelector = oneKeyWallet.NetworkSelector;
exports.ThirdwebNftMedia = oneKeyWallet.ThirdwebNftMedia;
exports.ThirdwebProvider = oneKeyWallet.ThirdwebProvider;
exports.Web3Button = oneKeyWallet.Web3Button;
exports.defaultTokens = oneKeyWallet.defaultTokens;
exports.oneKeyWallet = oneKeyWallet.oneKeyWallet;
exports.useBloctoWallet = oneKeyWallet.useBloctoWallet;
exports.useCoinbaseWallet = oneKeyWallet.useCoinbaseWallet;
exports.useEmbeddedWallet = oneKeyWallet.useEmbeddedWallet;
exports.useEmbeddedWalletSendVerificationEmail = oneKeyWallet.useEmbeddedWalletSendVerificationEmail;
exports.useEmbeddedWalletUserEmail = oneKeyWallet.useEmbeddedWalletUserEmail;
exports.useFrameWallet = oneKeyWallet.useFrameWallet;
exports.useInstalledWallets = oneKeyWallet.useInstalledWallets;
exports.useMetamask = oneKeyWallet.useMetamask;
exports.usePaperWallet = oneKeyWallet.usePaperWallet;
exports.usePaperWalletUserEmail = oneKeyWallet.usePaperWalletUserEmail;
exports.useRainbowWallet = oneKeyWallet.useRainbowWallet;
exports.useResolvedMediaType = oneKeyWallet.useResolvedMediaType;
exports.useShowConnectEmbed = oneKeyWallet.useShowConnectEmbed;
exports.useTrustWallet = oneKeyWallet.useTrustWallet;
exports.useWalletConnect = oneKeyWallet.useWalletConnect;
exports.useWalletConnectV1 = oneKeyWallet.useWalletConnectV1;
Object.defineProperty(exports, 'BloctoWallet', {
  enumerable: true,
  get: function () { return wallets.BloctoWallet; }
});
Object.defineProperty(exports, 'Coin98Wallet', {
  enumerable: true,
  get: function () { return wallets.Coin98Wallet; }
});
Object.defineProperty(exports, 'CoinbaseWallet', {
  enumerable: true,
  get: function () { return wallets.CoinbaseWallet; }
});
Object.defineProperty(exports, 'CoreWallet', {
  enumerable: true,
  get: function () { return wallets.CoreWallet; }
});
Object.defineProperty(exports, 'CryptoDefiWallet', {
  enumerable: true,
  get: function () { return wallets.CryptoDefiWallet; }
});
Object.defineProperty(exports, 'EmbeddedWallet', {
  enumerable: true,
  get: function () { return wallets.EmbeddedWallet; }
});
Object.defineProperty(exports, 'FrameWallet', {
  enumerable: true,
  get: function () { return wallets.FrameWallet; }
});
Object.defineProperty(exports, 'InjectedWallet', {
  enumerable: true,
  get: function () { return wallets.InjectedWallet; }
});
Object.defineProperty(exports, 'LocalWallet', {
  enumerable: true,
  get: function () { return wallets.LocalWallet; }
});
Object.defineProperty(exports, 'MagicLink', {
  enumerable: true,
  get: function () { return wallets.MagicLink; }
});
Object.defineProperty(exports, 'MetaMaskWallet', {
  enumerable: true,
  get: function () { return wallets.MetaMaskWallet; }
});
Object.defineProperty(exports, 'OKXWallet', {
  enumerable: true,
  get: function () { return wallets.OKXWallet; }
});
Object.defineProperty(exports, 'OneKeyWallet', {
  enumerable: true,
  get: function () { return wallets.OneKeyWallet; }
});
Object.defineProperty(exports, 'PaperWallet', {
  enumerable: true,
  get: function () { return wallets.PaperWallet; }
});
Object.defineProperty(exports, 'PhantomWallet', {
  enumerable: true,
  get: function () { return wallets.PhantomWallet; }
});
Object.defineProperty(exports, 'RabbyWallet', {
  enumerable: true,
  get: function () { return wallets.RabbyWallet; }
});
Object.defineProperty(exports, 'RainbowWallet', {
  enumerable: true,
  get: function () { return wallets.RainbowWallet; }
});
Object.defineProperty(exports, 'SafeWallet', {
  enumerable: true,
  get: function () { return wallets.SafeWallet; }
});
Object.defineProperty(exports, 'SignerWallet', {
  enumerable: true,
  get: function () { return wallets.SignerWallet; }
});
Object.defineProperty(exports, 'SmartWallet', {
  enumerable: true,
  get: function () { return wallets.SmartWallet; }
});
Object.defineProperty(exports, 'TrustWallet', {
  enumerable: true,
  get: function () { return wallets.TrustWallet; }
});
Object.defineProperty(exports, 'WalletConnect', {
  enumerable: true,
  get: function () { return wallets.WalletConnect; }
});
Object.defineProperty(exports, 'ZerionWallet', {
  enumerable: true,
  get: function () { return wallets.ZerionWallet; }
});
Object.defineProperty(exports, 'setWalletAnalyticsEnabled', {
  enumerable: true,
  get: function () { return wallets.setWalletAnalyticsEnabled; }
});
Object.keys(reactCore).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return reactCore[k]; }
  });
});
