// @thirdweb-dev/wallets/evm/wallets/signer/dist/thirdweb-dev-wallets-evm-wallets-signer.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var classPrivateFieldSet = require('../../../../dist/classPrivateFieldSet-f0dc2a42.cjs.dev.js');
var walletIds = require('../../../../dist/walletIds-d2cd7be8.cjs.dev.js');
var evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase = require('../../../../dist/base-94547c89.cjs.dev.js');
var chains = require('@thirdweb-dev/chains');
require('../../../../dist/classPrivateMethodGet-8801ecd5.cjs.dev.js');
require('../../abstract/dist/thirdweb-dev-wallets-evm-wallets-abstract.cjs.dev.js');
require('../../../../dist/defineProperty-982e7acf.cjs.dev.js');
require('ethers');
require('eventemitter3');
require('@thirdweb-dev/sdk');
require('../../../../dist/headers-3cc7b8c1.cjs.dev.js');

// eslint-disable-next-line @typescript-eslint/ban-types
var _storage = /*#__PURE__*/new WeakMap();
/**
 * @internal
 */
class SignerWallet extends evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase.AbstractClientWallet {
  constructor(options) {
    super("signerWallet", options);
    classPrivateFieldSet._classPrivateFieldInitSpec(this, _storage, {
      writable: true,
      value: void 0
    });
    if (options.clientId && options.chain) {
      options.chain = chains.updateChainRPCs(options.chain, options.clientId);
    }
    this.options = options;
    this.signer = options.signer;
    classPrivateFieldSet._classPrivateFieldSet(this, _storage, options?.storage || evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase.createAsyncLocalStorage(walletIds.walletIds.localWallet));
  }
  async getConnector() {
    if (!this.connector) {
      const {
        SignerConnector
      } = await Promise.resolve().then(function () { return require('../../../connectors/signer/dist/thirdweb-dev-wallets-evm-connectors-signer.cjs.dev.js'); });
      if (!this.signer) {
        this.signer = this.options.signer;
      }
      const defaults = (this.options.chain ? [...chains.defaultChains, this.options.chain] : chains.defaultChains).map(c => chains.updateChainRPCs(c, this.options.clientId));
      this.connector = new SignerConnector({
        chain: this.options.chain || chains.updateChainRPCs(chains.Ethereum, this.options.clientId),
        signer: this.signer,
        chains: this.chains || defaults,
        clientId: this.options.clientId,
        secretKey: this.options.secretKey
      });
    }
    return this.connector;
  }
}

exports.SignerWallet = SignerWallet;
